types/rlp_unmarshal.go

```diff
@@ -181,23 +218,50 @@ func (r *Receipts) UnmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) error {
 		return err
 	}
 
-	for _, elem := range elems {
-		rr := &Receipt{}
-		if err := rr.UnmarshalRLPFrom(p, elem); err != nil {
+	for i := 0; i < len(elems); i++ {
+		// Non-legacy tx raw contains a tx type prefix in the beginning according to EIP-2718.
+		// Here we check if the first element is a tx type and unmarshal it first.
+		txType := LegacyTx
+		if elems[i].Type() == fastrlp.TypeBytes {
+			if err = txType.UnmarshalRLPFrom(p, elems[i]); err != nil {
+				return err
+			}
+
+			// Then we increment element number in order to go to the actual tx data raw below.
+			i++
+		}
+
+		rr := &Receipt{
+			TransactionType: txType,
+		}
+
+		if err = rr.UnmarshalRLPFrom(p, elems[i]); err != nil {
 			return err
 		}
 
-		(*r) = append(*r, rr)
+		*r = append(*r, rr)
 	}
```

Comment:
Extract to helper function (same as with `Transaction` unmarshal logic)? The only line which differs between `rlp_unmarshal.go` and `rlp_unmarshal_storage.go` is this one: https://github.com/0xPolygon/polygon-edge/pull/1114/files#diff-885da5543ea77aa838b1210b3eb06bf214b06f2f77e0c2134051e4628efcbb15R152. WDYT?

Reply:
Done